psst-core/src/session/login5.rs

```diff
@@ -0,0 +1,269 @@
+// Ported from librespot
+
+use crate::error::Error;
+use crate::session::client_token::ClientTokenProvider;
+use crate::session::token::Token;
+use crate::session::SessionService;
+use crate::util::{default_ureq_agent_builder, solve_hash_cash};
+use librespot_protocol::login5::login_response::Response;
+use librespot_protocol::{
+    client_info::ClientInfo,
+    credentials::{StoredCredential},
+    hashcash::HashcashSolution,
+    login5::{
+        login_request::Login_method, ChallengeSolution, LoginError, LoginRequest,
+        LoginResponse,
+    },
+};
+use parking_lot::Mutex;
+use protobuf::well_known_types::duration::Duration as ProtoDuration;
+use protobuf::{Message, MessageField};
+use std::fmt::Formatter;
+use std::time::{Duration, Instant};
+use std::{error, fmt, thread};
+use crate::system_info::{CLIENT_ID, DEVICE_ID};
+
+const MAX_LOGIN_TRIES: u8 = 3;
+const LOGIN_TIMEOUT: Duration = Duration::from_secs(3);
```

Comment:
How did you determine these values? Testing or just adding to be cautious? They seem fine for now.